File: 2-Sorting/c++/5-merge-sort.cpp
#include <bits/stdc++.h>
using namespace std;

void merge(vector<int>& arr, int l, int mid, int r) {
    vector<int> left;
    vector<int> right;

    for (int i = l; i <= mid; i++) {
        left.push_back(arr[i]);
    }

    for (int i = mid + 1; i <= r; i++) {
        right.push_back(arr[i]);
    }

    int i = 0;
    int j = 0;

    int index = l;
    while (i < left.size() && j < right.size()) {
        if (left[i] < right[j]) {
            arr[index++] = left[i++];
        } else {
            arr[index++] = right[j++];
        }
    }

    while (i < left.size()) {
        arr[index++] = left[i++];
    }

    while (j < right.size()) {
        arr[index++] = right[j++];
    }
}

void merge_sort(vector<int>& arr, int l, int r) {
    if (r - l <= 0) {
        return;
    }

    int mid = (l + r) / 2;

    merge_sort(arr, l, mid);
    merge_sort(arr, mid + 1, r);

    merge(arr, l, mid, r);
}

void test() {
    vector<int> testVector({5, 1, 4, 7, 13, 2});
    vector<int> sortedVector({1, 2, 4, 5, 7, 13});

    merge_sort(testVector, 0, testVector.size() - 1);
    
    assert(testVector == sortedVector);
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    test();

    return 0;
}

File: 2-Sorting/rust/5-merge-sort.rs
// Merge sort

mod sort {
    /// Sort the given array via merge sort
    /// Complexity: O(array.len()*log2(array.len()))
    pub fn merge_sort(array: &mut Vec<i32>, left: i32, right: i32) {
        if left >= right {
            return;
        }
        
        let mid = (left + right) / 2;

        merge_sort(array, left, mid);
        merge_sort(array, mid + 1, right);
        
        merge(array, left as usize, mid as usize, right as usize);
    }
    
    fn merge(array: &mut Vec<i32>, left: usize, mid: usize, right: usize) {
        let left_half = array[left..mid + 1].to_vec();
        let right_half = array[mid + 1..right + 1].to_vec();
    
        let mut pos: usize = left;
        let mut i: usize = 0;
        let mut j: usize = 0;
    
        while i < left_half.len() && j < right_half.len() {
            if left_half[i] <= right_half[j] {
                array[pos] = left_half[i];
                i = i + 1;
            } else {
                array[pos] = right_half[j];
                j = j + 1;
            }
            pos = pos + 1;
        }
        
        while i < left_half.len() {
            array[pos] = left_half[i];
            i = i + 1;
            pos = pos + 1;
        }
        
        while j < right_half.len() {
            array[pos] = right_half[j];
            j = j + 1;
            pos = pos + 1;
        }
    }
}

#[cfg(test)]
mod tests {

    #[test]
    fn test_merge_sort() {
        let mut test_array: Vec<i32> = vec![5, 1, 4, 7, 13, 2];
        let expected_result: Vec<i32> = vec![1, 2, 4, 5, 7, 13];
    
        let from: i32 = 0;
        let to: i32 = (test_array.len() - 1) as i32;
        crate::sort::merge_sort(&mut test_array, from, to);
    
        assert_eq!(expected_result, test_array);
    }
}
